### ISSUE TITLE
How is versioned instrumentation chosen and enabled?

### ISSUE THREAD

**Issue by FrankSpitulski** [open-telemetry/opentelemetry-java-instrumentation#1065]
For example servlet 2 vs 3, the debug logs show both are loaded.

**Comment by trask** [comment]
During the build, we have a plugin (`muzzle`) that checks what library methods each instrumentation is utilizing, e.g. the `Servlet3Advice` calls `ServletRequest.isAsyncStarted()`, and that plugin encodes those method signatures into the instrumentations bytecode (you can see this if you decompile the Instrumentation classes :smile:). Then at runtime, the javaagent checks to see if the method signatures that instrumentation needs to use are present in the classloader, and only applies the instrumentation if they are all present (because otherwise that instrumentation would fail at runtime with [AbstractMethodError](https://docs.oracle.com/javase/7/docs/api/java/lang/AbstractMethodError.html):

> Thrown when an application tries to call an abstract method. Normally, this error is caught by the compiler; this error can only occur at run time if the definition of some class has incompatibly changed since the currently executing method was last compiled.

Documenting this in more detail is on my plate (#536), but also happy to answer more questions about it here.

**Comment by FrankSpitulski** [comment]
referring to tag 0.7.0 I found 
```java
  // this is required to make sure servlet 2 instrumentation won't apply to servlet 3
  @Override
  public ElementMatcher<ClassLoader> classLoaderMatcher() {
    // Optimization for expensive typeMatcher.
    return hasClassesNamed("javax.servlet.http.HttpServlet")
        .and(not(hasClassesNamed("javax.servlet.AsyncEvent", "javax.servlet.AsyncListener")));
  }
```
in `io.opentelemetry.auto.instrumentation.servlet.v2_2.Servlet2Instrumentation`.
Is this deprecated now and muzzle handles turning off servlet 2?

**Comment by trask** [comment]
Ah, no, this is still needed. The reason for

```
.and(not(hasClassesNamed("javax.servlet.AsyncEvent", "javax.servlet.AsyncListener")))
```

is that the Servlet 2 instrumentation would otherwise get applied to Servlet 3 (since all of the Servlet 2 methods still exist in Servlet 3).

**Comment by FrankSpitulski** [comment]
Is there any way to target different versions with different instrumentations where the versions have an identical api?

**Comment by trask** [comment]
There must be something different about the two versions?

**Comment by FrankSpitulski** [comment]
They wouldn't be different from the perspective of the instrumentation. I'll try a different scenario to better explain.

How could servlet 2 disable itself without knowing about servlet 3? If servlet 4 comes out with new apis and I add servlet 4 support in via otel.initializer.jar how could servlet 3 know to disable itself? In the same vein how could servlet 3 disable itself if I've added my own servlet 3 instrumentation (extending from the existing servlet 3 tracer) using otel.initializer.jar?

**Comment by trask** [comment]
> If servlet 4 comes out with new apis and I add servlet 4 support in via otel.initializer.jar how could servlet 3 know to disable itself?

it can't know to disable itself, but you can disable it via https://github.com/open-telemetry/opentelemetry-java-instrumentation#disabled-instrumentations

**Comment by FrankSpitulski** [comment]
Thanks 👍

**Comment by trask** [comment]
hey @FrankSpitulski, some of the instrumentation within a given module rely on each other, so the safest is to disable/enable at the module level.

can you explain your specific need?

**Comment by FrankSpitulski** [comment]
@trask, my use case is that I wanted to extend the Servlet3HttpServerTracer tracer with new tags and had to duplicate all of the servlet3 instrumentation. There were two distinct instrumentations in there the AsyncContextInstrumentation which I did not want to touch and the regular Servlet3Instrumentation both under the "servlet3" enable/disable tag. Ideally I could just swap out or extend the tracer, less ideally but still good I would only have to replace the Servlet3Instrumentation, but what I ended up having to do was duplicate both sets of instrumentation.

**Comment by trask** [comment]
Thanks. If we provide a hook for you to provide your own subclass of `Servlet3HttpServerTracer`, would that address your needs?

**Comment by FrankSpitulski** [comment]
Partially. For this specific use case yes. If I'm unable to get a workable solution for https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/894 then I'd have to change the instrumentation advice, not just the tracer. In that case I'd still like to be able to target specific instrumentation.

**Comment by iNikem** [comment]
@FrankSpitulski @trask this issue is labeled as "discussion". Do we still have something to discuss here or all raised asks are already addressed by #566 ? Can we close this issue?

**Comment by FrankSpitulski** [comment]
#566 doesn't address the ability to disable specific instrumentation. I found myself needing an extra argument while writing cassandra instrumentation. That wouldn't be possible with the hooks. In the cassandra case it's easy to just turn the whole thing off but if I wanted an extra argument in a servlet instrumentation I'd run into the same problem where I'd need to duplicate code rather than targeting a specific component of the servlet instrumentation to replace.

**Comment by iNikem** [comment]
Will be addressed by #566
